Allow for Dockerfile to be named something else.

duglin opened the pull request:
This is a follow-on to #7995 and #7284.

This PR allows for the Dockerfile to be named something other than `Dockerfile` via a -f/--file option. Aside from the name/path of the Dockerfile being variable, nothing else should change from today's behavior. The location of the Dockerfile MUST be within the build context.

This PR contains the:
- code changes
- docs updates
- testcases

Signed-off-by: Doug Davis dug@us.ibm.com

Commit:
Allow for Dockerfile to be named something else.
Add a check to make sure Dockerfile is in the build context
Add docs and a testcase
Make -f relative to current dir, not build context

Signed-off-by: Doug Davis <dug@us.ibm.com>

icecrime commented:
LGTM

Ping @jfrazelle @LK4D4: if one of you could please give it a last look and merge if it's ok!

tianon commented on api/client/commands.go:
```diff
@@ -139,10 +140,38 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
+			}
+		} else {
+			filename = path.Join(absRoot, *dockerfileName)
+		}
```
I'm a little confused as to why we've got this extra logic in here detecting whether `-f` or `--file` got passed in.

Couldn't we just change the default to be the empty string, and swap based on that to `CONTEXT/Dockerfile`?  I think it's going to be confusing for users if they see that the default for `-f` is `"Dockerfile"`, but if they explicitly do `-f Dockerfile` it's now actually semantically different, and means the `Dockerfile` in the current directory, regardless of the context.

duglin replied:
Before I make a change I want to get agreement on what I think you're asking for:
- change the default Dockerfile to be "CONTEXT/Dockerfile"  (keeps backwards compat)
- this will not only be a code change but a help-text change, so the end user will see:
         -f, --file="CONTEXT-ROOT/Dockerfile"    Name of the Dockerfile
- then if the user wants to be explicit and specify -f but still use the default value then they would need to include the path to the Dockerfile (including build context) on the -f arg.
Right?
Seems reasonable to me.

duglin replied:
Actually, thinking more about how the flag stuff works, I think the user might need to see:
-f, --file=""   Name of Dockerfile (default is CONTEXT-ROOT/Dockerfile)

tianon replied:
I meant change the default to the empty string, and use that to determine
when to change it to `<context-root>/Dockerfile`.  Text-parsing for
`CONTEXT/` or `CONTEXT-ROOT/` is going to be flaky IMO.  Then the text can
describe that the default is a file named `Dockerfile` at the root of the
context, but that an argument can be supplied which is either absolute or
relative to the CWD.

duglin replied:
Right, we're on the same page. Will do...

tianon commented on api/client/commands.go:
```diff
@@ -222,6 +251,9 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
 		v.Set("pull", "1")
 	}
+
+	v.Set("f", *dockerfileName)
```
Since we have the opportunity now before this merges and gets set in stone, shouldn't we give this API parameter a more verbose/meaningful name like `dockerfile`? :angel:

tianon replied:
The `dockerfileName` we use as the variable name everywhere sounds pretty good, too. :wink:

duglin replied:
@tianon which "API parameter" are you referring to ? The line of code this comment is on already uses "dockerfileName".

tianon replied:
I mean that "f" which is the API parameter.  A third-party API client
wanting to use this feature would have to specify a query parameter of
"f=..." to use this feature, which isn't very intuitive for an API IMO (but
maybe I'm alone in this?  we have an absolute plethora of similarly
badly-named API parameters already).

duglin replied:
Ah, well, the original PR I submitted used "-d" instead but at some point people seemed to gravitate towards -f (perhaps because 'make' uses -f to point to the makefile?) so I switched it.  I do think that if we use -d on the API then we should probably use it on the cmd line too. 
I'm ok with either, but would need the appropriate UX approval first I think.

tianon replied:
Nonono this string is for the API parameter, not the CLI.  I 100% agree
with "-f" for the CLI.  This line represents the _API_.

duglin replied:
ah ok but still the question is whether the API should match the CLI. Doesn't it usually?

tianon replied:
In some places, yes, in other places not so much.

duglin replied:
who gives the final decision on UX questions?

LK4D4 replied:
I'm +1 for changing name of this api parameter. All this `t`, `q` are pretty bad :(

duglin replied:
ok changed it to "dockerfile"

tianon commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -448,11 +448,12 @@ To kill the container, use `docker kill`.
-      -q, --quiet=false    Suppress the verbose output generated by the containers
-      --rm=true            Remove intermediate containers after a successful build
-      -t, --tag=""         Repository name (and optionally a tag) to be applied to the resulting image in case of success
+	  -f, --file="Dockerfile"  Location of the Dockerfile to use
```
Is this indented more than the following flags on purpose? :wink:

duglin replied:
there was a tab in there - converted it to spaces, so things should align now.

duglin commented:
@tianon I made the change to the default value of -f per your suggestion.

duglin commented:
@SvenDowideit I updated the API v1.17 docs - I missed adding "f" to the /build parameters. See if you're ok with the edits.

SvenDowideit commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -599,6 +606,28 @@ repository is used as Dockerfile. Note that you
+    $ sudo docker build -f Dockerfile.debug .
+
+This will use a file called `Dockerfile.debug` instead of `Dockerfile`
+during the build process. If a `Dockerfile` also exists it will be ignored.
```
perhaps it would be better not to use the word `ignored` - as `.dockerignore` means the file isn't added to the context?

maybe 

```
This will use a file called `Dockerfile.debug` for the build instructions, instead of the `Dockerfile`.
```

duglin replied:
done

SvenDowideit commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -599,6 +606,28 @@ repository is used as Dockerfile. Note that you
+
+The above commands will build the current build context twice, once
+using a debug version of a `Dockerfile` and once using a production
+version.
```
its worth re-enforcing that the context is build from the location of the final `.`

duglin replied:
Done

SvenDowideit commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -599,6 +606,28 @@ repository is used as Dockerfile. Note that you
+
+    $ cd /home/me/myapp/some/dir/really/deep
+    $ sudo docker build -f /home/me/myapp/dockerfiles/debug /home/me/myapp
+	$ sudo docker build -f ../../../../dockerfiles/debug /home/me/myapp
```
indenting?

SvenDowideit replied:
ooo, so if i grok right, the value of `-f` will be different to the `f=` value passed to the daemon - as the first is relative to the dir we're in (i get weirded out by this) and the second will be relative to the context?

duglin replied:
Fixed the indenting - had a tab in there by mistake.

re: -f vs f=  I added a bit more text under the -f/--file section to make it clear that if the value is a relative path then it must be relative to the current dir.

SvenDowideit commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -599,6 +606,28 @@ repository is used as Dockerfile. Note that you
+These two `docker build` commands do the exact same thing. They both
+use the contents of the `debug` file instead of looking for a `Dockerfile` 
+and will use `/home/me/myapp` as the root of the build context. Note that 
+debug is in the directory structure of the build context, regardless of how 
```
Note that the `debug` file must be within the directory...

duglin replied:
fixed

SvenDowideit commented on docs/man/docker-build.1.md:
```diff
@@ -30,6 +31,9 @@ When a Git repository is set as the **URL**, the repository is used
 
 # OPTIONS
+**-f**, **--file**=*Dockerfile*
+   Location of the Dockerfile to use. The file must be within the build context. The default is *Dockerfile*.
```
@duglin I don't think the docs here are clear on which it is - is `-f` relative to the `pwd`, or relative to the location of the context dir.

and as i note in the cli.md docs, if its relative to the `pwd` (which is how i read your docs atm), then the `-f` value will be different to the value sent to the daemon.

duglin replied:
ok I updated the docs here to make it clear that its relative to the current dir. See if it helps.

LK4D4 commented on api/client/commands.go:
```diff
@@ -139,10 +143,42 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
+			}
+		}
+
+		if !strings.HasSuffix(absRoot, "/") {
```
Why this checks done for default `Dockerfile` too? Shouldn't this be in non-default name condition?

LK4D4 replied:
Also, I am not sure, but probably better not do magic with trailing `/` and just use `filepath.Rel`

thaJeztah replied:
^^ also if this is on the client, it will probably break on Windows?

duglin replied:
ok I moved that logic to just be for user-defined paths and used filepath.Rel (didn't know that existed, cool!)

duglin replied:
Apparently filepath.Rel() will try really really hard to find a path that works - even if it has to generate one with ../../../.. to do it - notice the description even says:
On success, the returned path will always be relative to basepath, even if basepath and targpath *share no elements*

which doesn't make it nearly as useful.  Reverting parts of that change but open to other ideas.

LK4D4 commented on builder/evaluator.go:
```diff
@@ -136,7 +138,13 @@ func (b *Builder) Run(context io.Reader) (string, error) {
-	filename := path.Join(b.contextPath, "Dockerfile")
+	filename := path.Join(b.contextPath, b.dockerfileName)
+
+	absContextPath, _ := filepath.Abs(b.contextPath)
```
Just curious, why you not checking errors here? :) You checked them on client side.

duglin replied:
Not sure what I was thinking at the time other than it seemed pretty unlikely that they would fail but I'll add 'em.

SvenDowideit commented:
Docs LGTM - merci-much

LK4D4 commented:
@duglin need rebase :) sorry

duglin commented:
@LK4D4 ok rebased

erikh commented on api/client/commands.go:
```diff
@@ -136,10 +141,42 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
+
+		if *dockerfileName == "" {
+			// No -f/--file was specified so use the default
+			origDockerfile = "Dockerfile"
```
can we make this a constant?

erikh commented:
need the constant, otherwise LGTM

duglin commented:
@erikh ok created a constant and used it on client and server

duglin commented:
ok rebased and constant is made/used.

duglin commented:
no, I pushed it, but then #8748 caused a conflict.
rebased and added a testcase to make sure that the new .dockerignore processing (from #8748) works even when the Dockerfile is called something else.

crosbymichael commented:
Nice

crosbymichael commented:
Cool, i think this has enough LGTMs

dreamcat4 commented:
We also need this feature on the DockerHub (hub.docker.com) for it's Automated Builds. In the 'Edit Build Settings' page where it has the Field named 'Dockerfile Location'.

https://support.docker.com/hc/en-us/requests/3494
